clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,506 @@
+		Version:   "v1",
+		DataItems: npm.resultSummary.DataItems,
+	})
+	if err != nil {
+		klog.Info("Pretty Print to Json Err: ", err)
+	}
+	summaryName := fmt.Sprintf(npm.String() + npm.resultSummary.podRatio + npm.resultSummary.protocol + npm.resultSummary.service)
+	return measurement.CreateSummary(summaryName, "json", content), nil
+}
+
+// Dispose disposes resources,objects after the measurement.
+func (npm *networkPerformanceMeasurement) Dispose() {
+	if npm.framework == nil {
+		klog.V(1).Infof("Network measurement %s wasn't started, skipping the Dispose() step", npm)
+		return
+	}
+	klog.Infof("Stopping network performance measurement...")
+	close(npm.stopCh)
+	close(npm.podInfo.podPairCh)
+	npm.cleanupCluster()
+}
+
+func (npm *networkPerformanceMeasurement) handleUpdateNetworkTestEvents(newObj interface{}) {
+	newRuntimeObj, ok := newObj.(runtime.Object)
+	if newObj != nil && !ok {
+		klog.Errorf("error casting object: %s", newObj)
+		return
+	}
+
+	resourceContent, err := runtime.DefaultUnstructuredConverter.ToUnstructured(newRuntimeObj)
+	if err != nil {
+		klog.Errorf("error converting event Object to unstructured.Event object: %s", newObj)
+		return
+	}
+	var metricResp metricResponse
+	metricMap := resourceContent["status"].(map[string]interface{})
+	if err = getMetricResponse(metricMap, &metricResp); err != nil {
+		klog.Error(err)
+	}
+	if len(metricResp.Metrics) == 0 && metricResp.Error != "" {
+		klog.Error("error message from worker pod: ", metricResp.Error)
+	}
+	npm.metricLock.Lock()
+	npm.metricVal = append(npm.metricVal, metricResp.Metrics)
+	npm.metricLock.Unlock()
+}
+
+func (npm *networkPerformanceMeasurement) execNToMTest() {
+	go npm.createCustomResourcePerUniquePodPair()
+
+	npm.podInfo.formUniquePodPair()
+}
+
+func (npm *networkPerformanceMeasurement) createCustomResourcePerUniquePodPair() {
+	currTime := time.Now()
+	npm.startTimeStampForTestExecution = currTime.Add(initialDelayForTestExecution).Unix()
+	for {
+		select {
+		case uniqPodPair := <-npm.podInfo.podPairCh:
+			npm.templateMapping = npm.populateTemplate(uniqPodPair, uniqPodPair.podIndex)
+			if err := npm.framework.ApplyTemplatedManifests(networkTestRequestFilePath, npm.templateMapping); err != nil {
+				klog.Error(err)
+			}
+			npm.podPairNumber++
+		default:
+			return
+		}
+	}
+}
+
+func (wp *workerPodInfo) formUniquePodPair() {
+	for index := 0; len(wp.workerPodMap) > 0; index++ {
+		firstNode, sourcePod, err := wp.Pop()
+		destinationPod, err := wp.PopFromDifferentNodeThan(firstNode)
+		if err != nil {
+			break
+		}
+		uniquePodPair := uniquePodPair{
+			podIndex:           index,
+			sourcePodName:      sourcePod.podName,
+			sourcePodIP:        sourcePod.podIP,
+			destinationPodName: destinationPod.podName,
+			destinationPodIP:   destinationPod.podIP}
+		wp.podPairCh <- uniquePodPair
+	}
+}
+
+func (wp *workerPodInfo) Pop() (string, *workerPodData, error) {
+	for key, value := range wp.workerPodMap {
```

Comment:
Technically you should probably Pop() from the longest queue. Otherwise in cases like

1 pod on node A, 1 pod on node B, 2 pods on node C

you won't be able to create two pairs

Reply:
We have seen one issue in current implementation in comparison to earlier one. Earlier implementation iterated over all the map entries and popped one pod each from worker node. So at the end of one iteration of map, all worker nodes had equal size. With this implementation, we see sometimes the number of unique pod pairs are not as expected (With 10:10 ration i tested , i could see sometimes 9 pairs , 10 pairs some other time). We saw one worker node left with 2 pods which can't be paired with each other.

Reply:
Ya even i am wondering , how do we solve this problem of 2 pods on same node remaining without pairing. How do we know the longest queue in map. We can implement it , but a costly one in terms of performance . Everytime , we may have to iterate and figure out the longest queue.